pkg/mcs/resource_manager/server/manager.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/json"
+	"sort"
+	"sync"
+
+	"github.com/pingcap/errors"
+	rmpb "github.com/pingcap/kvproto/pkg/resource_manager"
+	"github.com/tikv/pd/server"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/storage"
+)
+
+// Manager is the manager of resource group.
+type Manager struct {
+	sync.RWMutex
+	groups  map[string]*ResourceGroup
+	storage func() storage.Storage
+	// TODO: dispatch resource group to storage node
+	getStores func() ([]*core.StoreInfo, error)
+}
+
+// NewManager returns a new Manager.
+func NewManager(srv *server.Server) *Manager {
+	getStores := func() ([]*core.StoreInfo, error) {
+		rc := srv.GetRaftCluster()
+		if rc == nil {
+			return nil, errors.New("RaftCluster is nil")
+		}
+		return rc.GetStores(), nil
+	}
+
+	m := &Manager{
+		groups:    make(map[string]*ResourceGroup),
+		storage:   srv.GetStorage,
+		getStores: getStores,
+	}
+	srv.AddStartCallback(m.Init)
+	return m
+}
+
+// Init initializes the resource group manager.
+func (m *Manager) Init() {
+	handler := func(k, v string) {
+		var group ResourceGroup
+		if err := json.Unmarshal([]byte(v), &group); err != nil {
+			panic(err)
+		}
+		m.groups[group.Name] = &group
+	}
+	m.storage().LoadResourceGroups(handler)
+}
+
+// AddResourceGroup puts a resource group.
+func (m *Manager) AddResourceGroup(group *ResourceGroup) error {
+	m.RLock()
+	_, ok := m.groups[group.Name]
+	m.RUnlock()
+	if ok {
+		return errors.New("this group already exists")
+	}
+	err := group.CheckAndInit()
+	if err != nil {
+		return err
+	}
+	if err := m.storage().SaveResourceGroup(group.Name, group); err != nil {
```

Comment:
There is window where write is lost here since no lock protection.

Reply:
storage should be thread-safe interface.

Reply:
I didn't get your point. How about following scenario:
2 client invoke AddResourceGroup concurrently with identical group name, but different attributes. Both can pass existing check and call m.storage().SaveResourceGroup(). The content of latter caller of request B is stored. But later it's possible that m.groups[] saves the content for request A in memory.

Reply:
make sense. updated.